### ISSUE TITLE
BLE support 

### ISSUE THREAD

**Issue by Ben1980** [libdivecomputer/libdivecomputer#32]
I'm a codeveloper on a flutter plugin that is wrapping [libdivecomputer](https://github.com/DiveNote/dive_computer)
I'm mocking the BLEObject implementation of [subsurface](https://github.com/subsurface/subsurface/blob/master/core/qt-ble.cpp) and therefore provided the iostream with a `dc_context_t` and vtable functions `set_timeout`, `poll`, `read`, `write`, `ioctl`, and `close`.
With the flutter plugin [flutter_blue_plus](https://github.com/boskokg/flutter_blue_plus) I get a working BLE connection to my Suunto EON Core. I then enable notifications for the characteristic and listen to the characteristics
```
Future<void> enableNotifications(int status) async {
    //Subsurface qt-ble.cpp line 650
    //Currently we ignore Heinrichs Weikamp dive computer

    if (prefeeredService == null) {
      status = dc_status_t.DC_STATUS_IO;
      return;
    }

    for (BluetoothCharacteristic characteristic
        in prefeeredService!.characteristics) {
      if (!_isReadCharacteristic(characteristic)) {
        continue;
      }

      final descriptors = characteristic.descriptors;
      BluetoothDescriptor clientDescriptor = descriptors.first;

      // Get client characteristic configuration descriptor
      for (BluetoothDescriptor descriptor in descriptors) {
        if (descriptor.uuid == Guid("00002902-0000-1000-8000-00805f9b34fb")) {
          clientDescriptor = descriptor;
          break;
        }
      }

      try {
        // writing 0x0100 to the client characteristic configuration descriptor
        // await clientDescriptor.write([0x0100]);
        log.info(
            'Set notify value for ${characteristic.characteristicUuid.toString()}');
        final result = await characteristic.setNotifyValue(true);
        status =
            result ? dc_status_t.DC_STATUS_SUCCESS : dc_status_t.DC_STATUS_IO;
      } catch (e, s) {
        status = dc_status_t.DC_STATUS_IO;
      }
    }
  }

Future<void> selectPreferredService(
      BluetoothDevice device, int status) async {
    //Subsurface qt-ble.cpp select_preferred_service
    .
    .
    .

    if (prefeeredService != null) {
      for (BluetoothCharacteristic characteristic
          in prefeeredService!.characteristics) {
        _characteristicSubscriptions
            .add(characteristic.lastValueStream.listen((value) {
          print(value);
          _characteristcStateChanged(characteristic, Uint8List.fromList(value));
        }));

        //print(_characteristicSubscriptions.length);
      }
    }
  }
```

The problem is now that the `lastValueStream` is triggering but the `value` is always 0. It seems that my Suunto is not broadcasting anything and I'm running into a timeout. 
I know that my code is based on flutter dart and libdivecomputer is in C which I try wrapping so it might be very difficult to get any help. But maybe you see why I'm doing wrong. For example I can't write ` 0x0100` to the configuration descriptor but as far as I know `setNotifications(true)` should do the same.

**Comment by jefdriesen** [comment]
Are you sure you are connecting to the correct service and characteristics? Usually there are several present, and you need to select the correct one. I maintain a list of the BLE service and characteristic for all the dive computers supported by libdivecomputer here:
https://github.com/libdivecomputer/libdivecomputer/wiki/BLE

**Comment by Ben1980** [comment]
Thanks, that's a good point. I will check if I have the right service and characteristics. I was mocking subsurface [select_preferred_service](https://github.com/subsurface/subsurface/blob/master/core/qt-ble.cpp#select_preferred_service) and I thought that actually would select the correct one.

**Comment by Ben1980** [comment]
Is this the normal behavior of the EON Core? Is the shown descriptor the one that will deliver the data? I'm not sure how this should work and appreciate any help.

**Comment by jefdriesen** [comment]
The service and characteristics shown in the screenshot are indeed the ones you should use.

**Comment by Ben1980** [comment]
Hm ok then i don't understand why my EON Core is not sending data. Anything else I could check?

**Comment by Ben1980** [comment]
Does anyone know how suuntos BLE interface is working or if there is some unofficial documentation available? I tried suuntos official support but they told me they don't support such requests because of security concerns 🤦‍♂️

**Comment by jefdriesen** [comment]
You'll only receive data **after** sending a (correct) command to the other `c6339440` characteristic. You need to enable notifications on the `d0fd6b80` characteristic to be able to receive data packets, but the dive computer doesn't send anything without first receiving a request. It's a request/response type of protocol.

The BLE service is simply an implementation of serial communication over a BLE link. One characteristic acts as the Rx line and the other as the Tx line.

**Comment by jefdriesen** [comment]
Are you able to download dives from your dive computer using a libdivecomputer based application (like Subsurface)? Just to rule out any problems with your system.

**Comment by Ben1980** [comment]
Yes it is working with subsurface mobile. I would debug it if I could get it working manually on my iPhone but this seems pretty complex.![image](https://github.com/libdivecomputer/libdivecomputer/assets/919260/dbd75ebf-9cb4-4740-a287-64504a40a42d)

**Comment by Ben1980** [comment]
Am I right that this is the code that is triggering download at ``?
```
static dc_status_t
suunto_eonsteel_send(suunto_eonsteel_device_t *device,
	unsigned short cmd,
	const unsigned char data[],
	unsigned int size)
{
	dc_status_t rc = DC_STATUS_SUCCESS;
	unsigned char buf[PACKET_SIZE + CRC_SIZE];

	// Two-byte packet header, followed by 12 bytes of extended header
	if (size + 2 + HEADER_SIZE + CRC_SIZE > sizeof(buf)) {
		ERROR(device->base.context, "Insufficient buffer space available.");
		return DC_STATUS_PROTOCOL;
	}

	memset(buf, 0, sizeof(buf));

	buf[0] = 0x3f;
	buf[1] = size + HEADER_SIZE;

	// 2-byte LE command word
	array_uint16_le_set(buf + 2, cmd);

	// 4-byte LE magic value (starts at 1)
	array_uint32_le_set(buf + 4, device->magic);

	// 2-byte LE sequence number;
	array_uint16_le_set(buf + 8, device->seq);

	// 4-byte LE length
	array_uint32_le_set(buf + 10, size);

	// .. followed by actual data
	if (size) {
		memcpy(buf + 14, data, size);
	}

	// 4 byte LE checksum
	unsigned int crc = checksum_crc32r(buf + 2, size + HEADER_SIZE);
	array_uint32_le_set(buf + 14 + size, crc);

	if (dc_iostream_get_transport(device->iostream) == DC_TRANSPORT_BLE) {
		rc = dc_iostream_write(device->iostream, buf + 2, size + HEADER_SIZE + CRC_SIZE, NULL);
	} else {
		rc = dc_iostream_write(device->iostream, buf, sizeof(buf) - CRC_SIZE, NULL);
	}
	if (rc != DC_STATUS_SUCCESS) {
		ERROR(device->base.context, "Failed to send the command.");
		return rc;
	}

	HEXDUMP (device->base.context, DC_LOGLEVEL_DEBUG, "cmd", buf + 2, size + HEADER_SIZE);

	return DC_STATUS_SUCCESS;
}
```

**Comment by jefdriesen** [comment]
Yes, that's the code that builds the packet and then sends it out.

Are you sure your data gets send out? Does your code even reaches the characteristic.write() call? I also don't see any error checking for the call, so maybe it fails? I don't remember if the Suunto characteristic is WriteWithoutResponse or WriteWithResponse, so maybe you are using the wrong type?

**Comment by jefdriesen** [comment]
Looking at the [flutter_blue_plus](https://github.com/boskokg/flutter_blue_plus?tab=readme-ov-file#subscribe-to-a-characteristic) documentation you have two options to subscribe: `onValueReceived` and `lastValueStream`. The latter also triggers for writes! That might be the problem. Your `enableNotifications` functions enables notifications for all READ characteristics, but your `selectPreferredService` function enables `characteristic.lastValueStream.listen` for ALL characteristics. Since that includes the WRITE characteristic, I guess that code will already trigger when you try to send a packet.

**Comment by Ben1980** [comment]
Yes, the `write()` function is called and writing without response leads to a success on `characteristic.write(...)`. I tried also false but then `characteristic.write(...)` is failing. So I guess the Suunto is accepting WriteWithNoResponse.
I have a suspicion that the conversion `Uint8List bytes = data.cast<ffi.Uint8>().asTypedList(size);` of the data sent to write() in dart is somehow wrong. Is there a way to check what the outcome should be? Currently, the resulting data looks like `[126, 0, 0, 1, 0, 0, 0, 0, 0, 4, 0, 0, 0, 2, 0, 42, 0, 122, 160, 156]` and `[48, 126]`.

**Comment by latishab** [comment]
> Yes, the `write()` function is called and writing without response leads to a success on `characteristic.write(...)`. I tried also false but then `characteristic.write(...)` is failing. So I guess the Suunto is accepting WriteWithNoResponse. I have a suspicion that the conversion `Uint8List bytes = data.cast<ffi.Uint8>().asTypedList(size);` of the data sent to write() in dart is somehow wrong. Is there a way to check what the outcome should be? Currently, the resulting data looks like `[126, 0, 0, 1, 0, 0, 0, 0, 0, 4, 0, 0, 0, 2, 0, 42, 0, 122, 160, 156]` and `[48, 126]`.

Yes, I have checked through LightBlue, the write characteristic is accepting .withoutResponse. I also can't get the response correctly through notifications. Please let me know if you have any update on this.

**Comment by jefdriesen** [comment]
The two data packets `[126, 0, 0, 1, 0, 0, 0, 0, 0, 4, 0, 0, 0, 2, 0, 42, 0, 122, 160, 156]` and `[48, 126]` are correct. That's the CMD_INIT command, which gets hldc encoded and then split into two BLE packets (of max 20 bytes). That's data that gets generated by libdivecomputer, so it's correct.

When you don't get a response, I see two possible reasons:
1. The above data packets are not getting send out for some reason. In that case the dive computer doesn't receive the packet, and thus won't respond either.
2. The notifications are not enabled correctly. Even if the dive computer receives the request and answers, your application doesn't get the response.

I checked some logs, and the c6339440 characteristic does indeed support WriteWithoutResponse.

**Comment by latishab** [comment]
> The two data packets `[126, 0, 0, 1, 0, 0, 0, 0, 0, 4, 0, 0, 0, 2, 0, 42, 0, 122, 160, 156]` and `[48, 126]` are correct. That's the CMD_INIT command, which gets hldc encoded and then split into two BLE packets (of max 20 bytes). That's data that gets generated by libdivecomputer, so it's correct.
> 
> When you don't get a response, I see two possible reasons:
> 
> 1. The above data packets are not getting send out for some reason. In that case the dive computer doesn't receive the packet, and thus won't respond either.
> 2. The notifications are not enabled correctly. Even if the dive computer receives the request and answers, your application doesn't get the response.
> 
> I checked some logs, and the c6339440 characteristic does indeed support WriteWithoutResponse.

Is suunto_eonsteel_device_open necessary before being able to send commands? Because I am trying to send commands using the suunto_eonsteel_send, after I connect with standard CoreBluetooth from iOS framework, the data sent through WriteWithoutResponse is successful but i'm not receiving anything from notifications.

**Comment by jefdriesen** [comment]
The `dc_iostream_t` interface is the low-level transport that is responsible for sending/receiving the data packets. The `dc_device_t` interface is the layer that is responsible for building those packets, together with the logic to download the dives. The latter uses the former to send/receive the data packets, and that's why you need to pass a `dc_iostream_t` instance as a parameter to the `suunto_eonsteel_device_open` function.

Technically you can use just the `dc_iostream_t` interface to communicate with the dive computer. In that case you'll need to manually build all those packets yourself. For a single test packet (to check whether your BLE code is working) that's certainly doable, but why would you do that if you can let libdivecomputer do that for you for free?

If you are not receiving any notifications, then I suspect you haven't enabled them correctly. Unfortunately I have no experience with CoreBluetooth.

**Comment by latishab** [comment]
> The `dc_iostream_t` interface is the low-level transport that is responsible for sending/receiving the data packets. The `dc_device_t` interface is the layer that is responsible for building those packets, together with the logic to download the dives. The latter uses the former to send/receive the data packets, and that's why you need to pass a `dc_iostream_t` instance as a parameter to the `suunto_eonsteel_device_open` function.
> 
> Technically you can use just the `dc_iostream_t` interface to communicate with the dive computer. In that case you'll need to manually build all those packets yourself. For a single test packet (to check whether your BLE code is working) that's certainly doable, but why would you do that if you can let libdivecomputer do that for you for free?
> 
> If you are not receiving any notifications, then I suspect you haven't enabled them correctly. Unfortunately I have no experience with CoreBluetooth.

Hi, i'm finally back in trying to resolve this myself with CoreBluetooth.
Here are the things that I have done (correct me if i'm wrong):
1. I pretty much setup functions in CoreBluetooth for connecting, services discovery, and characteristics discovery
2. Then i wrapped that up to a custom ble_object so that libdivecomputer able to use the io
3. I also created my functions for the callbacks through the dc_custom_cbs

Then I finally got this as a result:
DEBUG: Successfully connected to Suunto D5 204710004359
Discovered service: 98AE7120-E62E-11E3-BADD-0002A5D5C51B
Discovered characteristic: C6339440-E62E-11E3-A5B3-0002A5D5C51B
Write characteristic found
Discovered characteristic: D0FD6B80-E62E-11E3-A2E9-0002A5D5C51B
Notify characteristic found
Notification state updated: enabled
About to call dc_iostream_set_timeout
eon address: 0x301ce2760
eon->iostream address: 0x3009453e0
eon->iostream->vtable address: 0x1045dc738
eon->iostream->vtable->set_timeout address: 0x10459293c
Timeout value: 5000
iostream address: 0x3009453e0, in dc_hdlc_set_timeout
Timeout value: 5000
Returned from dc_iostream_set_timeout with status: 0
Error receiving notification: Reading is not permitted.
Received data from notify characteristic: 7e000202002df400003000000047555255330000
Received data from notify characteristic: 0000000000000000003230343731303030343335
Received data from notify characteristic: 390000000002059b074406000002059b07000400

Are the data received is correct? I basically converted it to a hexEncodedString
extension Data {
    func hexEncodedString() -> String {
        return map { String(format: "%02hhx", $0) }.joined()
    }
}

I'm sorry for asking you dumb questions.

**Comment by jefdriesen** [comment]
That looks like the first part of an HDLC frame:
```
7e000202002df400003000000047555255330000
0000000000000000003230343731303030343335
390000000002059b074406000002059b07000400
```
The HDLC start byte is 0x7E.
